src/Test/Spec/Reporter/TeamCity.purs

```diff
@@ -38,55 +33,104 @@ escape = Regex.replace'
-  renderProperty (key /\ value) = key <> "='" <> escape value <> "'"
+teamcity :: forall a. String -> ServiceMessage a -> String
+teamcity = teamcity' ""
+
+teamcity' :: forall a. String -> String -> ServiceMessage a -> String
+teamcity' rest event { name, nodeId, parentNodeId } = "##teamcity["
+  <> event
+  <> "name" := name
+  <> "nodeId" := nodeId
+  <> "parentNodeId" := fromMaybe "0" parentNodeId
+  <> rest
+  <> "]"
+
+property :: String -> String -> String
+property key value = " " <> key <> "='" <> escape value <> "'"
+
+infix 7 property as :=
+
+testCount :: Int -> String
+testCount count = "##teamcity[testCount count='" <> show count <> "']"
+
+testSuiteStarted :: forall a. ServiceMessage a -> String
+testSuiteStarted = teamcity' "" "testSuiteStarted"
+
+testSuiteFinished :: forall a. ServiceMessage a -> String
+testSuiteFinished = teamcity' "" "testSuiteFinished"
+
+testStarted :: forall a. ServiceMessage a -> String
+testStarted = teamcity' "" "testStarted"
+
+testIgnored :: forall a. ServiceMessage a -> String
+testIgnored = teamcity' "" "testIgnored"
+
+testFinished :: forall a. ServiceMessage a -> String
+testFinished = teamcity' "" "testFinished"
+
+testFinishedIn :: WithDuration -> String
+testFinishedIn d = teamcity' ("duration" := (show $ trunc d.duration)) "testFinished" d
+
+testFailed :: WithMessage -> String
+testFailed d = teamcity' ("message" := d.message) "testFailed" d
 
-testSuiteStarted :: Array (Tuple String String) -> String
-testSuiteStarted = teamcity "testSuiteStarted"
+type ServiceMessage x =
+  { name :: String
+  , nodeId :: String
+  , parentNodeId :: Maybe String
+  | x
+  }
 
-testSuiteFinished :: Array (Tuple String String) -> String
-testSuiteFinished = teamcity "testSuiteFinished"
+type WithMessage = ServiceMessage (message :: String)
+type WithDuration = ServiceMessage (duration :: Number)
 
-testStarted :: Array (Tuple String String) -> String
-testStarted = teamcity "testStarted"
+serviceMessage :: String -> Path -> ServiceMessage ()
+serviceMessage name path =
+  let
+    nodeId = idFromPath path
+    parentNodeId = parentSuite path
+      <#> _.path
+      <#> idFromPath
+  in
+    { name, nodeId, parentNodeId }
 
-testIgnored :: Array (Tuple String String) -> String
-testIgnored = teamcity "testIgnored"
+withDuration :: Number -> ServiceMessage () -> WithDuration
+withDuration duration { name, nodeId, parentNodeId } =
+  { name, nodeId, parentNodeId, duration }
 
-testFinished :: Array (Tuple String String) -> String
-testFinished = teamcity "testFinished"
+withMessage :: String -> ServiceMessage () -> WithMessage
+withMessage message { name, nodeId, parentNodeId } =
+  { name, nodeId, parentNodeId, message }
 
-testFailed :: Array (Tuple String String) -> String
-testFailed = teamcity "testFailed"
+idFromPath :: Path -> String
+idFromPath path = path
+  <#> unwrap
+  <#> (\{ index, name } -> show index <> ":" <> fromMaybe "" name)
```

Comment:
I wonder if name is really necessary here? From what I understand, `index` is global over the whole universe of tests.

Reply:
I added name to make the output easier to debug by humans. But your right its not necessary though the ":" is to differentiate "0:" from "0", since "0" is a special id that is only allowed as a `parentNodeId` for the top level messages.

Reply:
Yes, human debugging is a good point. I'm just a bit worried that test names may run afoul of TC's rules for IDs.

Reply:
In what way? that they are to long?

Reply:
That they may contain characters that TC doesn't like in an ID. Sorry, I only had cursory experience with TC, so I don't really know what rules are.

Reply:
The name is escaped according to spec (as far as I understood it).

Reply:
Oh I see. 
Ok, good enough for me.